### DISCUSSION TITLE
Possibility to setup charging / discharging time in Eco Mode

### DISCUSSION THREAD

**Discussion by ghost** [mletenay/home-assistant-goodwe-inverter#98] [Ideas]
Hello to all,

I saw new version of this integration and it is possible to switch between inverter modes (General mode, Off grid mode, Backup mode, Eco mode).
 
I have GoodWe 8K-ET inverter and 4x Risen/SYL HomeESS HV3700 (14,8kWh).

I would like to ask you, if is somehow possible to setup / change / add / enable - disable "Charging time for begging - end, percentage of current + modes charge / discharge and days of repeating" as is possible  in ECO Mode by using PV Master application (as is visible on picture which i attached).

Sorry picture is in Czech language, but i think you can see what i mean :
![PV_Master_screen_ECO_MODE](https://user-images.githubusercontent.com/10319031/147729868-b3a864aa-01e3-4a83-9ef2-eca73b08dba8.png)
)

I thing is the one strong and sens-full "UseCase" for this switching mode which is possible, but i don't find it in my Home Assistant.

I have done now some integration to read OTE prices and conversion into CZK prices. I have available some Weather prediction for next day by forecast. Those prices I get every day in evening for next day and i plan to make some calculation and make decision if and how many charge from grid i would like to get.

I have hourly tariff from Czech company www.bezdodavatele.cz when i can pay for power / sell over generation for market hourly prices and would be great, specially now when photovoltaic don't generate enough power in Winter, to make automation to buy and charge battery from grid. This I control today manually, but is boring every day setup it manual in PV Master.

I had plan to get command (in hex) for setup by listening of network communication between Mobile PV Master app and inverter with Wire shark, but first i would like to try ask if you have this done. Maybe is possible and i don't see only how.

![image](https://user-images.githubusercontent.com/10319031/147730271-dbbc7ad5-d203-4c58-a9d0-d9de2deaa8e9.png)

**Comment by mletenay** [comment]
No, at this moment support for the Eco mode is not implemented, neither at HomeAssistant nor in the underlying inverter library code.
In theory it could be done, I just never used that mode so had little interest to do it since it seems a bit more complex than rest of settings.
I'll try to look at it sometime this week/weekend.
If you want to help, I suggest the discord channel for faster communication.
(And yes, I understood the screenshot in czech, muzeme taky cesky/slovensky :-)

**Reply by ghost** [reply]
Yes, i would like to help, because i I have to work it out anyway. I thing it is better when i will extend your code, then i will start something own. When I have some time, I'll look in the GoodWe library and try to do some experiments. If I will be successful, we can look on it together and publish it in some of future version of your plugin.

I think this could be interesting for the others too.

I would like to join your discord channel and we can get in contact :-) can you send me a link to your channel?

**Reply by mletenay** [reply]
It's also in the project info - https://discord.gg/TaXyWXT

**Reply by mletenay** [reply]
I made a quick look on it yesterday and it seems to be straightforward
Those 4 mode on/off switches are modbus registers 0xB99E, B9A2, B9A6 and B9AA
To set the mode values it issues multi write command (0x10) to 3 registers to 0xB99B, B99F, B9A3 and B9A7
Values are startH, startM, endH, endM, power (2b), 255 a repeatDays

**Reply by mletenay** [reply]
Basic support for reading the Eco Mode group values was added here - https://github.com/marcelblijleven/goodwe/commit/3578ff0a2fbb47f297c22db9cc84255e67fe6c24

**Reply by ghost** [reply]
did you think about how to visualize it? I can try to use your prepared function and play with. Now it will be simple :-)

For some automation it is now great / enough, but i think it would be nice to get maybe some Lovelance widget for manual setting. Good starting point to learn this technology, because i am new in python / lovelance custom cards.

**Reply by jansemrad** [reply]
Hi I have almost same setup as @VitBarabas, now I would like to automate charging/discharging battery based on NodeRed automation. From NodeRed I am able to change switch general and eco mode, how to change the Eco mode setting values to start charging in eco mode. Sorry for may be basic question, but I sm new to HA.

**Reply by jansemrad** [reply]
Hi did you managed to implement the charging into the HA? In case yes please can you shar solution with me? Thank you.

**Comment by Jayksopp** [comment]
I'm also interested in automating charging during off peak hours when expected solar production is too low. Since 0.9.7 we can set eco mode properly. I will try to set eco mode to charge all the time in PV-master. That way I can automate switching between eco mode and general mode without setting the charging times in HA.

Would be nice to have a discharge setting somehow so we can force discharge to the grid when prices are high?

**Comment by tinuva** [comment]
@Jayksopp 

Why not set the inverter to backup mode? When I do that my ES inverter starts charging from the grid.

**Comment by jansemrad** [comment]
I have ET inverter, and switching to backup mode means that is not charging but as well not discharging, when connected to grid. So my scenario is as follows - charging from PV panels + discharging (peak hours) = general mode, charging (cheapest hours) = ECO mode + select profile inside eco mode (not possible in HA yet - only here:  https://github.com/mletenay/home-assistant-goodwe-inverter/discussions/98#discussioncomment-1908944, not discharging, only charging from PV panels between Eco and General mode. This I calculated based on historical prices and my consumption is the best scenario. This I tested with my invertor and this is working. There is only one thing, when I fully charge the batteries with ECO mode and switch to General mode with HA, it don´t want to start discharging, until I finish CT test (this is possible to activate in last step of mode selection wizard).

**Reply by mletenay** [reply]
Just from curiosity, why you want/need the CT test ? 
I was under impression it is used/useful only during installation, to confirm all the clamps are installed on right phases and in right direction.
I expect no other usage for the CT test functionality during "normal" operation, am I wrong ?

**Comment by jansemrad** [comment]
My expectation was the same, but if I charge the batteries with eco-mode and switch back to General mode by HA, batteries are not giving back any single Watt to the home (they are fully charged, but they are not discharging) . The only solution is to perform CT test, and then they start to operate normally.

**Reply by mletenay** [reply]
So it may be HA work mode switch problem (something is missing there).
If you do that "normally" via PvMaster is the CT test required too ?

**Reply by fojtp** [reply]
Why do you switch between eco mode and general mode?  If you turn off time schedules in eco mode, it behaves the same as general mode.  After switching the mode, the control mode runs for about 10 minutes during which the battery is not discharged.

**Reply by Jayksopp** [reply]
Ah interesting. But how do you turn off/change the eco mode time schedule from HA?

**Reply by jansemrad** [reply]
Good question, now I setup that ECO mode schedules with PV Master (I don´t heave yet implemented them in HA - don´t have time for it) and back to General mode I am switching with HA, that was pretty easy to implement in Node Red. I will test your solution.

**Reply by jansemrad** [reply]
> Ah interesting. But how do you turn off/change the eco mode time schedule from HA?

As I wrote. Now I am turning on the Eco mode schedule from PV master manually (start charging), and turning of by enabling General mode in HA/Node red automatically (stop charging) - this is already implemented in HA. But the turn on Eco mode schedule need some python programing and I don´t have time/enough know how for it right now.

**Reply by fojtp** [reply]
One more thing - after the FW upgrade, a toggle switch in PV master (Advanced settings) appeared to turn on the battery charging (with the SOC option).  The disadvantage is that the charging current cannot be set and therefore it is charged with the maximum current according to BMS (in my case 25A).

**Reply by Jayksopp** [reply]
I did some tests and can confirm that for my inverter (GW3648-EM) if I switch to eco mode in PV master and then switch back to general mode from HA the inverter works normally and I don't need to do the CT test.

I still can't get the eco mode from HA to work. I'm pretty sure switching to general mode (both in HA and in PV-master) overwrites the settings in eco mode to 00:00:00 and disable all 4 groups. 
If I switch back to eco mode I see the switch is working as expected but the time settings etc are no longer there. 
So it just behaves as if it were general mode. 

It seems eco mode and general mode are the same but eco mode has time settings and general does not. The invertor is not storing the eco mode settings locally. We'll have to send them every time we change.

For now no way to automate charging without writing settings to eco mode as well.

**Reply by jansemrad** [reply]
I can confirm the same on GW10K-ET, just finished the test, switching form charging in Eco Mode to General mode over PV Master switch immediately into discharging mode, everything is working as expected.

**Reply by mletenay** [reply]
OK, so the CT test is not actually necessary, just some action invoked during that test has to be invoked also when switching from Eco mode and that one is missing in HA code, right ?

**Reply by Jayksopp** [reply]
No switching from eco mode to general is working now. It's only when switching back that we miss the group, time, days and power limit settings. 
If you can make a "charge mode" that would work? It can be switch to eco mode but with the settings included. I would go for charge every day all the time with some default power limit. 
Same could be done for "discharge mode". 
Then we can fully automate charge and discharge from HA.

**Reply by jansemrad** [reply]
Now I am confused and I don´t have answer for you. I started the charging procedure from PV Master (switching Eco mode + schedule) and after approximately 10 minutes of charging switched back to General mode with HA, and everything is OK. It means it started to deliver electricity from batteries to home. Twice before I must run the CT test... There was only one difference, now I was charging almost full battery for 10 minutes  in previous cases (where I have to start the CT test) it was from 10% to full. May be I can perform more tests for you, could you advice me how to easily setup ECO-mode + schedule from HA? That I can implement them into my charging scenario.

**Reply by jansemrad** [reply]
> No switching from eco mode to general is working now. It's only when switching back that we miss the group, time, days and power limit settings. If you can make a "charge mode" that would work? It can be switch to eco mode but with the settings included. I would go for charge every day all the time with some default power limit. Same could be done for "discharge mode". Then we can fully automate charge and discharge from HA.

This is the best solution as well from my perspective!
Setting two different ECO modes for Charging / Discharging = 00:00 - 23:59,  every day (and this is as well answer on my previous post how to easily implement charging into HA)

**Reply by mletenay** [reply]
Trouble is that the UI part is quite complex, res. has too many values.
What I can try to do in the short term is to expose HA service which will setup/turnOn/Off the eco group by taking properly formatted string parameter, something like "13:30-14:40 Mon,Wed,Thu -60% On".
And such string could be created/controlled by HA template stuff.
Would that help ?

**Reply by jansemrad** [reply]
Yes, I thing this could help or as I wrote before in my post. Everybody will control just charge / discharge and the times/dates will be controlled by HA. So the easiest way would by add to modes (General, Backup, OffGrid, Eco) two additional Charge and Discharge.
Charge: "00:00-23:59 Mon,Tue,Wed,Thu,Fri,Sat,Sun  -100% On"
Discharge: vice versa

**Reply by mletenay** [reply]
Aaah, I finally understood. So inverter would not be the one watching date/time, this would be delegated to HA and the inverter will just turn it on/off until instructed otherwise.
Hmm, makes sense and that should be easy to implement.

**Reply by jansemrad** [reply]
Exactly, I don´t expect, that somebody with HA will control the date/time with inverter, when he has all control data inside HA including Battery SoC etc. So this is quick win and easy to use solution for everybody. Thank you!!!! You are doing great job!

**Reply by ghost** [reply]
> Aaah, I finally understood. So inverter would not be the one watching date/time, this would be delegated to HA and the inverter will just turn it on/off until instructed otherwise. Hmm, makes sense and that should be easy to implement.

Don’t forget to add there numbers of group otherwise you can’t make more records in one day or over two days...

**Reply by mletenay** [reply]
See comment below - v0.9.8

**Comment by eriklefevre** [comment]
I have the same inverter. 5 kw version
Before Goodwe customer support installed a new firmware on my inverter I was also loosing the timer settings in the eco mode. Power group 1
I think before I could not use group 1 and 2 but 3 and 4 was working ok when switching from general to eco.
I still have group 4 in use and this is working fine.
Firmware is 1010E now.
Hope this helps.

***@***.***

***@***.***


From: jansemrad ***@***.***>
Reply to: mletenay/home-assistant-goodwe-inverter ***@***.***>
Date: Monday, 10 January 2022 at 18:11
To: mletenay/home-assistant-goodwe-inverter ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [mletenay/home-assistant-goodwe-inverter] Possibility to setup charging / discharging time in Eco Mode (Discussion #98)


Now I am confused and I don´t have answer for you. I started the charging procedure from PV Master (switching Eco mode + schedule) and after approximately 10 minutes of charging switched back to General mode with HA, and everything is OK. It means it started to deliver electricity from batteries to home. Twice before I must run the CT test... There was only one difference, now I was charging almost full battery for 10 minutes in previous cases (where I have to start the CT test) it was from 10% to full. May be I can perform more tests for you, could you advice me how to easily setup ECO-mode + schedule from HA? That I can implement them into my charging scenario.

—
Reply to this email directly, view it on GitHub<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmletenay%2Fhome-assistant-goodwe-inverter%2Fdiscussions%2F98%23discussioncomment-1941731&data=04%7C01%7C%7C0969f9c1e5524ba94ef808d9d45c4a1d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637774315106587910%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=6z%2BOTlPbzLGqdbK4%2Fa5BEGpxeNtTGQW%2FTi3Q3aTIz0Q%3D&reserved=0>, or unsubscribe<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAQ7BMDNJ4EZE3D2EHGY4BKTUVMHNJANCNFSM5K7H6KMQ&data=04%7C01%7C%7C0969f9c1e5524ba94ef808d9d45c4a1d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637774315106587910%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=ZDNGIB8pbHYnrGXoBz9ad%2B8oVLkeMW7wjzNMHSxxBbU%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7C%7C0969f9c1e5524ba94ef808d9d45c4a1d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637774315106587910%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=79LT0fBoVjJsjNzwsF7junnYXzflND4XhXXKay5tQs0%3D&reserved=0> or Android<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7C%7C0969f9c1e5524ba94ef808d9d45c4a1d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637774315106587910%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=WEBg89nuZOo54q5c0AZalOlbxGqE0LP%2Bvclb9Qzjifw%3D&reserved=0>.

**Comment by mletenay** [comment]
So. I've just released v0.9.8.
There are 2 new operation modes - Eco charge mode and Eco discharge mode.
There's also new number entity - Eco mode power.
Selection of one of these new (fake) modes should switch inverter to Eco Mode with single charging/discharging group valid 24/7 with charging/discharging power taken from the "Eco mode power" entity.
(Changing Eco mode power value is not automatically reflected, it is read and applied only during mode change at this moment).

Happy testing, it should work on both ET and ES inverters.

**Reply by jansemrad** [reply]
I just tested it and there is still some small bug see attached log when selecting Eco charge/discharge mode (I think conversion str  > int/float for eco_mode_power is missing) :-):

Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/goodwe/select.py:83
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: 06:09:16 (3 occurrences)
Last logged: 06:09:59

[140629419598320] bad operand type for abs(): 'str'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1495, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1530, in _execute_service
    await handler.job.target(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 209, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 663, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 896, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 700, in _handle_entity_call
    await result
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 54, in async_select_option
    await entity.async_select_option(option)
  File "/config/custom_components/goodwe/select.py", line 83, in async_select_option
    await self._inverter.set_operation_mode(operation_mode, eco_mode_power)
  File "/usr/local/lib/python3.9/site-packages/goodwe/et.py", line 345, in set_operation_mode
    await self.write_setting('eco_mode_1', EcoMode("1", 0, "").encode_charge(eco_mode_power))
  File "/usr/local/lib/python3.9/site-packages/goodwe/sensor.py", line 247, in encode_charge
    return bytes.fromhex("0000173b{:04x}ff7f".format((-1 * abs(eco_mode_power)) & (2 ** 16 - 1)))
TypeError: bad operand type for abs(): 'str'

**Comment by Jayksopp** [comment]
same for me:
![image](https://user-images.githubusercontent.com/96485220/148917533-b60f0212-9ab1-4baf-9b24-8095615cdf4c.png)

**Reply by mletenay** [reply]
Damn ;-(
Last moment changes .... My bad.

**Reply by mletenay** [reply]
I did quick fix, please try the master version

**Reply by jansemrad** [reply]
Please give me quick advice how to download new master branch to HA. I am not familiar with Git. I redownload the library, from Hacs, but there is still same error.

**Reply by tinuva** [reply]
> Please give me quick advice how to download new master branch to HA. I am not familiar with Git. I redownload the library, from Hacs, but there is still same error.

Go to HACS
Then to integrations.

You will see 3 dots , click on that and see the menu:
![Screenshot 2022-01-11 at 12 42 13](https://user-images.githubusercontent.com/5417682/148928377-b34107fb-5eaf-4b84-95d3-1e9d9e0a3e5a.png)

In the next window, select "master" as the version, then download
![Screenshot 2022-01-11 at 12 43 37](https://user-images.githubusercontent.com/5417682/148928403-4871a367-502c-4cc7-b7a5-ec831a673e85.png)

Home assistant will require a restart to take affect.

**Reply by jansemrad** [reply]
Redownloaded master, still same error:

Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/goodwe/select.py:83
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: 11:55:59 (2 occurrences)
Last logged: 11:56:15

[140648857584640] bad operand type for abs(): 'str'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1495, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1530, in _execute_service
    await handler.job.target(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 209, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 663, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 896, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 700, in _handle_entity_call
    await result
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 54, in async_select_option
    await entity.async_select_option(option)
  File "/config/custom_components/goodwe/select.py", line 83, in async_select_option
    await self._inverter.set_operation_mode(operation_mode, eco_mode_power)
  File "/usr/local/lib/python3.9/site-packages/goodwe/et.py", line 345, in set_operation_mode
    await self.write_setting('eco_mode_1', EcoMode("1", 0, "").encode_charge(eco_mode_power))
  File "/usr/local/lib/python3.9/site-packages/goodwe/sensor.py", line 247, in encode_charge
    return bytes.fromhex("0000173b{:04x}ff7f".format((-1 * abs(eco_mode_power)) & (2 ** 16 - 1)))
TypeError: bad operand type for abs(): 'str'

**Reply by Jayksopp** [reply]
Same for me

**Reply by jansemrad** [reply]
I think i should be may be like this:?

encode_charge
return bytes.fromhex("0000173b{:04x}ff7f".format((-1 * abs(int(eco_mode_power))) & (2 ** 16 - 1)))

**Reply by mletenay** [reply]
I don't understand what's going on there. 
The eco_mode_power IS integer.
Can you check in HA's developer tools section if you have entity with name "number.eco_mode_power" ?

**Reply by mletenay** [reply]
Or can you re-download the master again ? 
HACS seems to have issues these days, it's slow and unreliable. Wait 10s or more to master is displayed in combo and download after and restart HA.

**Reply by jansemrad** [reply]
This is exactly what I was doing twice, waiting about 50 seconds to show versions, then selected master and restarted HA after download was finished.

**Reply by jansemrad** [reply]
Just to be on the safe side, I redownloaded it again and is still the same.

**Reply by Bojkas1985** [reply]
Same problem here. I tried resetting the number to an integer without a decimal point, but the same error.
![image](https://user-images.githubusercontent.com/41238303/148945806-4a358119-e577-47aa-bbc0-2815041c7dc4.png)

**Reply by mletenay** [reply]
Try latest master, it's fixed there

**Comment by jansemrad** [comment]
Sorry is in Czech, but it's selfexplanatory:
![image](https://user-images.githubusercontent.com/95675224/148939940-b7543af8-94a2-4004-ada8-645b7433a60f.png)

**Reply by mletenay** [reply]
Aaargh, the state is float and my python skills are still limitted :-(
Pushed another hotfix, re-download the master and hope it works now.
(I will have to redesign it diferrently in the evening ...)

**Reply by jansemrad** [reply]
Still the same:
Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/goodwe/select.py:83
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: 14:00:20 (1 occurrences)
Last logged: 14:00:20

[139954794627232] bad operand type for abs(): 'str'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1495, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1530, in _execute_service
    await handler.job.target(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 209, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 663, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 896, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 700, in _handle_entity_call
    await result
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 54, in async_select_option
    await entity.async_select_option(option)
  File "/config/custom_components/goodwe/select.py", line 83, in async_select_option
    await self._inverter.set_operation_mode(operation_mode, eco_mode_power)
  File "/usr/local/lib/python3.9/site-packages/goodwe/et.py", line 345, in set_operation_mode
    await self.write_setting('eco_mode_1', EcoMode("1", 0, "").encode_charge(eco_mode_power))
  File "/usr/local/lib/python3.9/site-packages/goodwe/sensor.py", line 247, in encode_charge
    return bytes.fromhex("0000173b{:04x}ff7f".format((-1 * abs(eco_mode_power)) & (2 ** 16 - 1)))
TypeError: bad operand type for abs(): 'str'

**Reply by mletenay** [reply]
Nope, can't be.
After first "fix" there was different error, you still see the first one.
Don't know why, but it is not the master you're testing.

**Reply by Bojkas1985** [reply]
@jansemrad I guess I don't have to tell you, but did you restart? :)

**Reply by Bojkas1985** [reply]
@mletenay It probably won't be, but can't it be the fact that I've already used all 4 slots? Otherwise, I still have the same problem even though I installed the master version of the add-on.
`[1866379520] bad operand type for abs(): 'str'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1495, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1530, in _execute_service
    await handler.job.target(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 209, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 663, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 896, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 700, in _handle_entity_call
    await result
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 54, in async_select_option
    await entity.async_select_option(option)
  File "/config/custom_components/goodwe/select.py", line 83, in async_select_option
    await self._inverter.set_operation_mode(operation_mode, eco_mode_power)
  File "/usr/local/lib/python3.9/site-packages/goodwe/et.py", line 345, in set_operation_mode
    await self.write_setting('eco_mode_1', EcoMode("1", 0, "").encode_charge(eco_mode_power))
  File "/usr/local/lib/python3.9/site-packages/goodwe/sensor.py", line 247, in encode_charge
    return bytes.fromhex("0000173b{:04x}ff7f".format((-1 * abs(eco_mode_power)) & (2 ** 16 - 1)))
TypeError: bad operand type for abs(): 'str'`

**Reply by jansemrad** [reply]
Sorry my/git fault when it appears in the selection box versions and you select master, you have to wait until it changes from grey to white (approx 15 sec.)... Then is selected master branch for download. My apology. I started the charging mode and invertor is charging, now I will test switching between modes.

**Reply by mletenay** [reply]
Happened to me too, HACS is awfully slow recently.

**Reply by Bojkas1985** [reply]
Ok, the jansemrad solution helped. By default, I have the ECO mode set all the time, if I switch from Eco charge mode back to ECO, the charging time will remain active. True, I will no longer need to have permanent ECO and I can switch it back to General automatically.
It would definitely be great if you could set the time, but it's still an amazing job and that's exactly what I've been waiting for! I have already started thinking about replacing WIFI with LAN and using modbus.

**Reply by jansemrad** [reply]
So result after some quick tests. ECO charging, is charging and the regulation of current is working. If you select ECO Discharge mode with 100% value it starts to discharge with full current to my home and the rest is going to Grid. Warning this could be not expected behavior. I thing the correct solution for ECO Discharge should be just deactivate group in ECO mode.
P.S. Switching to General mode from ECO charging works as expected.

**Reply by Jayksopp** [reply]
Can confirm for me everything is working as well w latest master version! Switch to eco charge works. Switch back to general mode works as well. 

For me the eco discharge behavior is expected. This is useful for exporting power to grid when prices are high.

**Reply by jansemrad** [reply]
> Can confirm for me everything is working as well w latest master version! Switch to eco charge works. Switch back to general mode works as well.
> 
> For me the eco discharge behavior is expected. This is useful for exporting power to grid when prices are high.

You are right - it's understandable, but it was little surprise for me, that I can as well export from batteries. It should be good described, to prevent surprises. ;-)

**Reply by Bojkas1985** [reply]
OT: Do any of you use the OTE Rate application from Grinco? It is a very interesting application for hourly market prices, unfortunately only for today. Prices are always updated at 1 pm, but I won't find out until midnight. Does anyone have solved it somehow?
https://github.com/grinco/ote_rate

**Reply by jansemrad** [reply]
Well, agree with you using OTE Rate, I was as well thinking about it, that it can optimize consumption/charging based on tomorrows price. But don´t have solution for it yet.

**Reply by mletenay** [reply]
@Bojkas1985 Not sure if it still is the case, but there were reports by one or two users that reliability of the LAN module is actually worse than WIFI. Don't know details or if it was related to specific firmware version, just to let you know ...

**Reply by jansemrad** [reply]
Yes you are right, in the firmware version of the GoodWe LAN module V1.0.14 was big problem and it was not possible to communicate over UDP directly neither with PV Master nor HA. After update to V1.0.15k is everything reliable for 100%.

**Comment by mletenay** [comment]
V0.9.8.1 was just released.
It contains few fixes and improves behavior of the Eco mode power number entity (it reacts immediately, as one would expect).

**Reply by jansemrad** [reply]
Thank you, it's excelent. I am already running automatic charging with node red scenario. May be one think, after every install the number.eco_mode_power reverts to 0. Would it be possible to keep previously set walue or may be make as default 100%?

**Comment by SnokxStefan** [comment]
Hi, total newby regarding home assistant. I manage to get the Hacs Goodwe (experimental) to work. I can change operation mode by selecting the mode in the available selector. However, when i set the state using pyscript, the operation mode shows as set, however the inverter does not change mode. can you please give me an example in pyscript on how to change the operation mode using a pyscript ?
